Lombiq.HelpfulLibraries/Libraries/Mvc/TypedRoute.cs

```diff
@@ -15,17 +17,24 @@ namespace Lombiq.HelpfulLibraries.Libraries.Mvc
 {
     public class TypedRoute

```

Comment:
On a second thought, I'd approach this differently, by channeling typed route declarations to `UrlHelper.Action()`. For example, the URL can be prefixed not just by the tenant name, which is rather tenant URL prefix, but also by a virtual directory; and the prefix can be present also for the Default tenant. This is something that the class is not handling. Similarly, it's missing that routes can be declared not just as attributes, or that the Orchard dashboard's URL prefix is configurable My point is not that routing is complex and can be affected by a lot of dynamic things. Trying to build an URL corresponding to a controller action by hand is futile.

This also makes caching unnecessary (unless you can produce a cache key for an expression much cheaper than it is to just process the expression) since that's handled by the routing infrastructure.

Reply:
The problem is that we don't have access to the UrlHelper from the UI test context which was the primary motivator to create this class. Also there are UrlHelper and LinkGenerator, I'm pretty sure if I need to support one I need to support the other too. Both require a live HTTP context which you don't have during UI testing nor in a bunch of other scenarios.
One option is to add two methods that take IUrlHelper and LinkGenerator respectively as alternatives to ToString which I think is suitable for most UI Testing use cases.

Reply:
After implementing it, I think just the `LinkGenerator` method is enough. That goes into the `HttpContext` extension method that the `IOrchardHelper` uses. Can't make an extension method directly for the `@Url` because we need to inject the `ITypeFeatureProvider` since that's the fastest and most reliable way to retrieve the method name. (I guess it could be an argument on the extension method, but that feels too cumbersome.)